### ISSUE TITLE
Removing single container from deployment makes updating impossible

### ISSUE THREAD

**Issue by dwangfaulkner-fn** [pulumi/pulumi-kubernetes#2336]
### What happened?

I have a Kubernetes deployment managed by Pulumi. I would like to remove a sidecar container from the deployment, but I receive the error. I am using server-side apply and do not want to downgrade to client-side apply.

```
    error: Preview failed: 1 error occurred:
    	* the Kubernetes API server reported that "default/example-db" failed to fully initialize or become live: Deployment.apps "example-db" is invalid: spec.template.spec.containers[1].image: Required value
```

The diff is as follows:
```
         self.sample_db_deployment = kubernetes.apps.v1.Deployment(
             "example-db",
             opts=opts,
             api_version="apps/v1",
             kind="Deployment",
            metadata=kubernetes.meta.v1.ObjectMetaArgs(
                name="example-db",
                namespace="default",
                annotations={"pulumi.com/patchForce": "true"},
            ),
            spec=kubernetes.apps.v1.DeploymentSpecArgs(
                 template=kubernetes.core.v1.PodTemplateSpecArgs(
                    spec=kubernetes.core.v1.PodSpecArgs(
                        containers=[
                            kubernetes.core.v1.ContainerArgs(
                                image=image,
                                name="example-db",
                                // ...
                             ),
-                            kubernetes.core.v1.ContainerArgs(
-                                image=sidecar_image,
-                                name="sidecar",
-                              // ...
                         ],
                         init_containers=[
                             kubernetes.core.v1.ContainerArgs(
```

### Expected Behavior

If I remove a sidecar container from my Pulumi deployment spec, the deployment is updated and new pods are created without the sidecar container.

### Steps to reproduce

Create a Deployment with a spec that contains two containers. Remove the second container.

### Output of `pulumi about`

```
CLI          
Version      3.57.1
Go Version   go1.20.2
Go Compiler  gc

Plugins
NAME        VERSION
kubernetes  3.24.1
python      unknown

Host     
OS       darwin
Version  13.2.1
Arch     x86_64

This project is written in python: executable='/Users/david.wangfaulkner/deploy/pulumi/.venv/bin/python3' version='3.10.4
'

Current Stack: dev

# Omitting our stack output

Found no pending operations associated with dev

Backend        
Name           MBP-David-WangFaulkner.local
URL            gs://freenome-services-dev-pulumi-state
User           david.wangfaulkner
Organizations  

Pulumi locates its logs in /var/folders/j0/gwtznfkd1492ml31_l8z0y200000gp/T/ by default
warning: Failed to get information about the Pulumi program's dependencies: illegal semver returned by language host: types-PyYAML@6.0.12.8: Invalid character(s) found in patch number "12.8"
```

Not sure why it can't parse PyYAML? Here's my pyproject.toml dependencies:
```
[tool.poetry.dependencies]
python = "^3.9"
click = "==8.1.3"
pulumi = "==3.53.1"
pulumi-kubernetes = "==3.24.1"
pyYAML = "==6.0"
pytest-asyncio = "==0.20.3"
colorama = "0.4.6"
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by squaremo** [comment]
This looks related to the bug fixed in https://github.com/pulumi/pulumi-kubernetes/pull/2332, which will make its way into the next release.

**Comment by aureq** [comment]
@dwangfaulkner-fn we shipped v3.24.2 last week, could you please confirm this is now working on your end please?

**Comment by phillipedwards** [comment]
I spent a bit more time looking into this one, from a repro standpoint and it seems behavior has slightly changed from OP's experience. I end up getting the following error on the initial `pulumi up` after importing the deployment.

`od nginx-8597b79d67-nqmzp]: containers with unready status: [sidecar] -- [CrashLoopBackOff] back-off 5m0s restarting failed container=sidecar pod=nginx-8597b79d67-nqmzp_default(7a38a108-1141-4cd0-b143-b1f661a90816)`

Steps to repro are:
- Bring a cluster
- `kubectl apply -f deploy.yaml` where deploy.yaml contains the deployment yaml from above.
- `pulumi import kubernetes:apps/v1:Deployment deployment nginx --provider urn:pulumi:phil::aws-ts-eks::pulumi:providers:kubernetes::provider`
- remove the sidecar from the deployment, in code, and `pulumi up`
- See the above CrashLoopBackOff error
- Re run `pulumi up` and problem is now gone.

**Comment by rquitales** [comment]
Looked into this again, and it appears that there is a deeper bug when switching from CSA -> SSA with arrays and diffing. Thankfully, this will be resolved in the upcoming v4 release of this provider.

I've added a new test case in #2472 to ensure that this is working as expected. When this test is run against our v3 provider, it fails, but passes on the v4 branch. We likely won't be back porting a fix to v3 as the upcoming v4 release contains a bunch of server side related changes.

Some additional notes on https://github.com/pulumi/pulumi-kubernetes/issues/2336#issuecomment-1544011951:

-  The sidecar container isn't actually removed if you inspect the kubernetes deployment with `kubectl`, so it's still broken despite the the successful `pulumi up` operation. This is not an issue in the v4 branch.
- The CrashLoopBackOff error is actually due to the deployment configuration itself, and not from Pulumi or Kubernetes. When we define 2 nginx containers in a pod, they both attempt to create a server on the default ports (eg. `:80`). This will cause one of the nginx containers to fail due to a port collision as the other nginx container has already spun up. The test I added circumvents this by having the sidecar container be a generic ubuntu container that just sleeps. Alternatively, we could also have updated the nginx configuration to use a different port.
